### PR TITLE
Document fuzzing strategies

### PR DESCRIPTION
**Pull request by mkaput**

### CODE REVIEW

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Strategies
+
+While a felt is a _lingua franca_ of Cairo runtime, its whole value space is rarely covered by
+domain space in tested code.
+
+Using the [`given`](../02-cheatcodes/given.md) cheatcode,
+one can instruct Protostar how constrained or adjusted the value space of fuzzed inputs should be.
+
+By default, Protostar applies the [`strategy.unsigned()`](#strategyunsigned) strategy to all felt
+parameters.
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/03-fuzzing/strategies.md:10]: Line breaks could be removed.

Screen Shot 2022-08-04 at 17 59 08

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    %}
+    assert_le(a, b)
+    return ()
+end
+```
+
+## Performance caveats
+
+While it is possible to call `given()` multiple times in a test case,
+it is recommended to always do it once in a test case,
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:32]: `in a test case` is repeated

```suggestion
While it is possible to call `given()` multiple times in a test case,
it is recommended to always do it once,
```

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+and configure multiple input parameters by passing many arguments to the cheatcode.
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:33]: It could be a sentence.

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Due to Cairo language limitations and resulting therefrom implementation details,
+each call to the `given` cheatcode causes the fuzzer to restart immediately with fresh state
+and newly set strategies,
+with the maximum number of tested examples reduced by the amount used in previous runs.
+Think of this as Protostar _learning a new strategy_ (this terminology is used internally).
+The consequence of this process is that each next _learning step_ effectively reduces a space
+of explored input parameters values, thus making the test less effective.
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:41]: The most important information should be at the top.  This and the paragraph above could be summarized into two sentences:

> It's not recommended to call `given` cheatcode multiple times in a test case. Each call reduces a space of explored input parameter values making the test less effective.

https://developers.google.com/tech-writing/one/paragraphs#answer_what_why_and_how

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
+# Strategies
+
+While a felt is a _lingua franca_ of Cairo runtime, its whole value space is rarely covered by
+domain space in tested code.
```
Comment (kasperski95) [inline website/docs/tutorials/07-testing/03-fuzzing/strategies.md:4]: Can you dumb it down? (lingua franca, domain)

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 1 above)
Comment (kasperski95) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:10]: This section doesn't answer the question: "What is a Strategy?" (and how is a strategy related to the `given` cheatcode).

Comment (mkaput) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:10]: I don't agree that line breaks could be removed. These are separate paragraphs that say independent things.

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 5 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:4]: Done.

File: website/docs/tutorials/07-testing/03-fuzzing/strategies.md (hunk 1 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/03-fuzzing/strategies.md:10]: Done.

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 4 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:41]: I have left more of this paragraph, because you suggestion oversimplified things as it's actually important to say that the fuzzer starts from the beginning

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+While it is possible to call `given` multiple times in a test case, try to always do it once.
+Multiple input parameters should be configured by passing many arguments to the cheatcode.
+Due to Cairo language limitations and resulting therefrom implementation details,
+each call to the `given` cheatcode causes the fuzzer to restart immediately with a fresh state.
+In consequence, each call effectively reduces a space of explored input parameters values,
+thus making the test less effective.
+
+Avoid binding fuzzing strategies to input parameters values or any other external state,
+such as random data of system time.
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:39]: Maybe simpler formulation would be:

> Best practice is to hardcode fuzzing strategies parameters .

Or something similar

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    return ()
+end
+```
+
+## Performance caveats
+
+While it is possible to call `given` multiple times in a test case, try to always do it once.
+Multiple input parameters should be configured by passing many arguments to the cheatcode.
+Due to Cairo language limitations and resulting therefrom implementation details,
+each call to the `given` cheatcode causes the fuzzer to restart immediately with a fresh state.
```
Comment (MaksymilianDemitraszek) [inline website/docs/tutorials/07-testing/02-cheatcodes/given.md:34]: Is the information about fuzzer restarting necessary here 🤔

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 4 above)
Comment (kasperski95) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md]: You could've explained why a call reduces a space of input parameters below, though.
> It's not recommended to call given cheatcode multiple times in a test case. Each call reduces a space of explored input parameter values making the test less effective.
> ...DETAILS

https://untools.co/minto-pyramid

**Review by kasperski95 — APPROVED** [review]

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 7 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:34]: It is, because this has many consequences:
1. Random seed will be restarted.
2. Fuzzer will explore the same values once again.
3. If we add #499 then it may happen that fuzzer will test these examples again also

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 6 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:39]: doesn't this slightly change the meaning from "you shouldn't do that" to "it's best that you do the opposite"? I would prefer the "you shouldn't do that" voice.

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 4 above)
Comment (mkaput) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md]: now is it better? :)

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 7 above)
Comment (kasperski95) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:34]: Maybe these consequences should be in the doc.

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 4 above)
Comment (kasperski95) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md]: I phrased the remark incorrectly. It was meant to fix the oversimplification in the example in my comment.
Anyway, I've already approved this PR. Resolving.

File: website/docs/tutorials/07-testing/02-cheatcodes/given.md (hunk 7 above)
Comment (MaksymilianDemitraszek) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:34]: Will user know about this test being performed? Or It will be under the hood?

Comment (mkaput) [reply website/docs/tutorials/07-testing/02-cheatcodes/given.md:34]: User will know. These tests subtract max_examples, so that sum of all examples over all learning steps is = max_examples (sans shrinking phase)

**Review by MaksymilianDemitraszek — APPROVED** [review]
